bee-ledger/src/workers/storage.rs

```diff
@@ -0,0 +1,496 @@
+    batch: &mut <B as BatchBuilder>::Batch,
+    output_id: &OutputId,
+    output: &ConsumedOutput,
+) -> Result<(), Error> {
+    Batch::<OutputId, ConsumedOutput>::batch_insert(storage, batch, output_id, output)
+        .map_err(|e| Error::Storage(Box::new(e)))?;
+    Batch::<Unspent, ()>::batch_delete(storage, batch, &(*output_id).into())
+        .map_err(|e| Error::Storage(Box::new(e)))?;
+
+    Ok(())
+}
+
+pub(crate) fn delete_consumed_output_batch<B: StorageBackend>(
+    storage: &B,
+    batch: &mut <B as BatchBuilder>::Batch,
+    output_id: &OutputId,
+) -> Result<(), Error> {
+    Batch::<OutputId, ConsumedOutput>::batch_delete(storage, batch, output_id)
+        .map_err(|e| Error::Storage(Box::new(e)))?;
+    Batch::<Unspent, ()>::batch_insert(storage, batch, &(*output_id).into(), &())
+        .map_err(|e| Error::Storage(Box::new(e)))?;
+
+    Ok(())
+}
+
+pub(crate) async fn apply_balance_diffs<B: StorageBackend>(
+    storage: &B,
+    balance_diffs: &BalanceDiffs,
+) -> Result<(), Error> {
+    let mut batch = B::batch_begin();
+
+    apply_balance_diffs_batch(storage, &mut batch, balance_diffs).await?;
+
+    storage
+        .batch_commit(batch, true)
+        .await
+        .map_err(|e| Error::Storage(Box::new(e)))
+}
+
+pub(crate) async fn apply_balance_diffs_batch<B: StorageBackend>(
+    storage: &B,
+    batch: &mut <B as BatchBuilder>::Batch,
+    balance_diffs: &BalanceDiffs,
+) -> Result<(), Error> {
+    for (address, diff) in balance_diffs.iter() {
+        let balance = fetch_balance_or_default(storage, address).await?.apply_diff(diff)?;
+
+        if balance.amount() != 0 {
+            Batch::<Address, Balance>::batch_insert(storage, batch, address, &balance)
+                .map_err(|e| Error::Storage(Box::new(e)))?;
+        } else {
+            Batch::<Address, Balance>::batch_delete(storage, batch, address)
+                .map_err(|e| Error::Storage(Box::new(e)))?;
+        }
+    }
+
+    Ok(())
+}
+
+pub(crate) async fn apply_milestone<B: StorageBackend>(
+    storage: &B,
+    index: MilestoneIndex,
+    created_outputs: &HashMap<OutputId, CreatedOutput>,
+    consumed_outputs: &HashMap<OutputId, (CreatedOutput, ConsumedOutput)>,
+    balance_diffs: &BalanceDiffs,
+    migration: &Option<Migration>,
+) -> Result<(), Error> {
+    let mut batch = B::batch_begin();
+
+    let mut created_output_ids = Vec::with_capacity(created_outputs.len());
+    let mut consumed_output_ids = Vec::with_capacity(consumed_outputs.len());
+
+    insert_ledger_index_batch(storage, &mut batch, &index.into())?;
+
+    for (output_id, output) in created_outputs.iter() {
+        insert_created_output_batch(storage, &mut batch, output_id, output)?;
+        created_output_ids.push(*output_id);
+    }
+
+    for (output_id, (_, consumed_output)) in consumed_outputs.iter() {
+        insert_consumed_output_batch(storage, &mut batch, output_id, consumed_output)?;
+        consumed_output_ids.push(*output_id);
+    }
```

Comment:
```rust
let created_output_ids = created_outputs.iter()
    .map(|(output_id, output)| {
        insert_created_output_batch(storage, &mut batch, output_id, output)?;
        *output_id)
    })
    .collect::<Vec<_>>();
```

Ditto with `consumed_output_ids`.

Death to mutability! (also means you don't need to think about the `with_capacity`)

Reply:
The `?` is not working here

Reply:
What error did you get? It should do.

Reply:
Was lacking some time annotation